README.md
# NAME

Sys::Stats - Multi-platform, pure perl, system statistics gathering.

# VERSION

This document describes Sys::Stats version 0.01

# SYNOPSIS

    use Sys::Stats;
    
    my $ss = Sys::Stats->new();
    my %stats = $ss->all;

# DESCRIPTION

Sys::Stats is my attempt at having an all-in-one statistics gatherer.
There are many like this one, but this one is mine... well not really.

In order to make this simple I've stolen... hummm, borrowed rather, the
way the Munin project gather data on a host and made it a pure perl
implementation of it (barring slurping `/proc` files and
`system`-ing `/sbin` utilities).

So far, this module will only work on Linux and FreeBSD systems.
Hopefully more drivers will come in the future, oh and Patches welcomed.

# TODO

Those would be nice to have.

- Add more documentation

    List all the methods available and document their returned values.

- Add more drivers

    (Open|Net)BSD and Mac OSX would be nice.

- Add some tests.

    How do you tests ever changing values?

# LICENSE

Copyright (C) Geraud CONTINSOUZAS.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Geraud CONTINSOUZAS &lt;gcs@cpan.org>


Changes
Revision history for Perl extension Sys-Stats

{{$NEXT}}

0.01 2015-09-14T03:10:43Z

    - original version



lib/Sys/Stats.pm
package Sys::Stats;
use 5.010001;
use strict;
use warnings;

use Carp ();

our $VERSION = "0.01";

sub new {
    my $driver = 'Sys::Stats::' . ucfirst($^O);
    eval "require $driver";
    if ($@) {
        Carp::croak "Can't find a proper class for OS $^O : $@";
    }
    return bless {}, $driver;
}

sub all {
    my $self = shift;
    return (
            disk_usage   => [$self->disk_usage],
            inode_usage  => [$self->inode_usage],
            load_average => [$self->loadavg],
            if_stat      => [$self->if_stat],
            netstat      => {$self->netstat},
            memory       => {$self->memory},
        );
}

1;
__END__

=encoding utf-8

=head1 NAME

Sys::Stats - Multi-platform, pure perl, system statistics gathering.

=head1 VERSION

This document describes Sys::Stats version 0.01

=head1 SYNOPSIS

    use Sys::Stats;
    
    my $ss = Sys::Stats->new();
    my %stats = $ss->all;

=head1 DESCRIPTION

Sys::Stats is my attempt at having an all-in-one statistics gatherer.
There are many like this one, but this one is mine... well not really.

In order to make this simple I've stolen... hummm, borrowed rather, the
way the Munin project gather data on a host and made it a pure perl
implementation of it (barring slurping C<< /proc >> files and
C<< system >>-ing C<< /sbin >> utilities).

So far, this module will only work on Linux and FreeBSD systems.
Hopefully more drivers will come in the future, oh and Patches welcomed.

=head1 TODO

Those would be nice to have.

=over 4

=item Add more documentation

List all the methods available and document their returned values.

=item Add more drivers

(Open|Net)BSD and Mac OSX would be nice.

=item Add some tests.

How do you tests ever changing values?

=back

=head1 LICENSE

Copyright (C) Geraud CONTINSOUZAS.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Geraud CONTINSOUZAS E<lt>gcs@cpan.orgE<gt>

=cut

# vim: syn=perl nu ai cin ts=4 et sw=4 fdm=marker


lib/Sys/Stats/Linux.pm
package Sys::Stats::Linux;
use 5.010001;
use strict;
use warnings;

use Capture::Tiny   'capture';
use File::Slurper   'read_text';
use File::Spec;
use List::MoreUtils 'firstval';

use parent 'Sys::Stats';

our $VERSION = "0.01";

sub _df_cmd { # {{{
    my $cmd  = '/bin/df';
    my @args = @_;
    my @exclude_fs
        = map { "-x ${_}" }
            qw(none unknown rootfs iso9660 squashfs udf romfs ramfs debugfs cgroup_root devtmpfs); 
    push @args, @exclude_fs;
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    shift @lines;

    my @res;
    foreach my $line (@lines) {
        my ($fs, $total, $used, $avail, undef, $mnt) = split(/\s+/, $line, 6);
        $fs =~ s/[^a-zA-Z0-9_]/_/g;
        push @res, +{
                filesystem  => $fs,
                total       => $total,
                used        => $used,
                available   => $avail,
                mount_point => $mnt,
            };
    }
    return @res;
} # }}}

sub disk_usage { return _df_cmd(qw(-P -l)); }

sub inode_usage { return _df_cmd('-i'); }

sub loadavg { # {{{
    my $file = '/proc/loadavg';
    Carp::croak "File '$file' is not readable"
        unless (-r $file);
    my ($avg1, $avg5, $avg15) = split /\s+/, read_text($file);
    return ($avg1, $avg5, $avg15);
} # }}}

sub list_nics { # {{{
    my $file = '/proc/net/dev';
    Carp::croak "File '$file' is not readable"
        unless (-r $file);

    return grep { $_ ne '' }
        map { /^\s*(?<interface>[^:]+):/ ? $+{interface} : '' }
            grep { ! /^\s*(lo|sit\d+):/ }
                split($/, read_text($file));
} # }}}

sub if_stat { # {{{
    my @res;
    foreach my $nic (list_nics()) {
        my $path = "/sys/class/net/$nic/statistics";

        my %stats
            = map {
                    chomp(my $value = read_text(File::Spec->catfile($path, $_)));
                    ($_ => $value)
                }
                qw(rx_bytes rx_packets tx_bytes tx_packets);
        push @res, +{ interface => $nic, %stats };
    }
    return @res;
} # }}}

sub netstat { # {{{
    my $cmd  = '/bin/netstat';
    my @args = qw(-s);
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    # FIXME 20150910 geraud
    # Thank you! https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=541172
    if ($exit) {
        my $ret_code = $exit >> 8;
        Carp::croak "Cannot call '$cmd @args' : $err"
            if ($ret_code > 1);
    }

    my @lines = split $/, $out;
    my %res_value_for = (
            'active connections ope'  => "active",
            'passive connection ope'  => "passive",
            'failed connection'       => "failed",
            'connection resets'       => "resets",
            'connections established' => "established",
        );
    my %res;
    foreach my $rx (keys %res_value_for) {
        my $line = firstval { /$rx/ } @lines;
        if ($line) {
            $line =~ s/^\s*//;
            my ($value) = split(/\s+/, $line);
            $res{$res_value_for{$rx}} = $value;
        }
    }
    return %res;
} # }}}

sub memory { # {{{
    my $file = '/proc/meminfo';
    Carp::croak "File '$file' is not readable"
        unless (-r $file);
    
    my %meminfo
        = map {
            my ($key, $val) = split /\s*:\s*/, $_;
            my ($v, $unit)  = split /\s+/, $val;
            $unit //= '';
            $v *= ($unit eq 'mB') ? 1024*1024
                : ($unit eq 'kB') ? 1024
                :                   1;
            ($key => $v);
        }
            split $/, read_text($file);

    my $shmem      = $meminfo{Shmem}     // 0;
    my $cached     = $meminfo{Cached}    // 0;
    my $swap_free  = $meminfo{SwapFree}  // 0;
    my $swap_total = $meminfo{SwapTotal} // 0;
    return (
            total    => $meminfo{MemTotal} // 0,
            active   => $meminfo{Active}   // 0,
            inactive => $meminfo{Inactive} // 0,
            cached   => $cached - $shmem,
            free     => $meminfo{MemFree}  // 0,
            swap     => $swap_total - $swap_free,
            buffers  => $meminfo{Buffers}  // 0,
        );
} # }}}

1;
__END__

=encoding utf-8

=head1 NAME

Sys::Stats::Linux - Linux driver for Sys::Stats.

=head1 VERSION

This document describes Sys::Stats::Linux version 0.01

=head1 DESCRIPTION

You shouldn't use this module directly. Please refer to L<Sys::Stats> instead.

=head1 LICENSE

Copyright (C) Geraud CONTINSOUZAS.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Geraud CONTINSOUZAS E<lt>gcs@cpan.orgE<gt>

=cut

# vim: syn=perl nu ai cin ts=4 et sw=4 fdm=marker


lib/Sys/Stats/Freebsd.pm
package Sys::Stats::Freebsd;
use 5.010001;
use strict;
use warnings;

use Capture::Tiny   'capture';
use List::MoreUtils 'firstval';

use parent 'Sys::Stats';

our $VERSION = "0.01";

sub disk_usage { # {{{
    my $cmd  = '/bin/df';
    my @args = (
            '-P',
            '-k', 
            '-t', 
            join(',',qw(noprocfs devfs fdescfs linprocfs linsysfs nfs nullfs cd9660))
        );

    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    shift @lines;

    my @res;
    foreach my $line (@lines) {
        my ($fs, $total, $used, $avail, undef, $mnt) = split(/\s+/, $line, 6);
        $fs =~ s/[^a-zA-Z0-9_]/_/g;
        push @res, +{
                filesystem  => $fs,
                total       => $total,
                used        => $used,
                available   => $avail,
                mount_point => $mnt,
            };
    }
    return @res;
} # }}}

sub inode_usage { # {{{
    my $cmd  = '/bin/df';
    my @args = (
            '-i',
            '-t', 
            join(',',qw(noprocfs devfs fdescfs linprocfs linsysfs nfs nullfs cd9660))
        );
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    shift @lines;

    my @res;
    foreach my $line (@lines) {
        my ($fs, undef, undef, undef, undef, $used, $avail, undef, $mnt) = split(/\s+/, $line, 9);
        $fs =~ s/[^a-zA-Z0-9_]/_/g;
        push @res, +{
                filesystem  => $fs,
                total       => $used + $avail,
                used        => $used,
                available   => $avail,
                mount_point => $mnt,
            };
    }
    return @res;
} # }}}

sub loadavg { # {{{
    my $cmd  = '/sbin/sysctl';
    my @args = qw(-n vm.loadavg);
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my (undef, $avg1, $avg5, $avg15) = split /\s+/, $out;
    return ($avg1, $avg5, $avg15);
} # }}}

sub list_nics { # {{{
    my $cmd = '/usr/bin/netstat';
    my @args = qw(-i -b -n);
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    shift @lines;
    my @res;
    foreach my $line (@lines) {
        next if ( $line =~ m{^(faith|lo\d|pflog)} );
        if ( $line =~ m{<Link#[0-9]*>} ) {
            $line =~ s/\** .*//;
            push @res, $line;
        }
    }
    return @res;
} # }}}

sub if_stat { # {{{
    my $cmd = '/usr/bin/netstat';
    my @args = qw(-i -b -n);
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    shift @lines;
    my @res;
    foreach my $line (@lines) {
        # Ignore pseudo interfaces
        next if ( $line =~ m{^(faith|lo\d|pflog)} );

        if ( $line =~ m{<Link#[0-9]*>} ) {
            my @e = split /\s+/, $line;
            # Down interface has a '*' marker appended to it. Remove it.
            $e[0] =~ s/\**$//;

            # KLUDGE 20150910 geraud
            # The munin code says (in awk style)
            # if (NF == 10) { 
            #     rbytes = $6; obytes = $9;
            # } else {
            #     rbytes = $7; obytes = $10;
            # }
            # Problem is, on my box, NF is either 11 or 12, and I can't find a
            # netstat(1) with a 10 columns output (oldest available is 8.4). 
            # For now, adjust accordingly.
            my $base_idx
                = (@e == 11) ?  3 
                : (@e == 12) ?  4
                :               0;
            next unless $base_idx;

            push @res, +{
                interface =>  $e[0],
                rx_packets => $e[$base_idx],
                rx_bytes  =>  $e[$base_idx + 3],
                tx_packets => $e[$base_idx + 4],
                tx_bytes  =>  $e[$base_idx + 6],
            };
        }
    }
    return @res;
} # }}}

sub netstat { # {{{
    my $cmd  = '/usr/bin/netstat';
    my @args = qw(-s);
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    my %res_value_for = (
            'connection requests'     => "active",
            'connection accepts'      => "passive",
            'bad connection'          => "failed",
            'reset$'                  => "resets",
            'connections established' => "established",
        );
    my %res;
    foreach my $rx (keys %res_value_for) {
        my $line = firstval { /$rx/ } @lines;
        if ($line) {
            $line =~ s/^\s*//;
            my ($value) = split(/\s+/, $line);
            $res{$res_value_for{$rx}} = $value;
        }
    }
    return %res;
} # }}}

sub _swapinfo { # {{{
    my $cmd  = '/usr/sbin/swapinfo';
    my @args = qw(-k);
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    shift @lines;
    my @res;
    foreach my $line (@lines) {
        my ($device, $total, $used, $avail, undef) = split /\s+/, $line;
        push @res, +{
                device => $device,
                total  => $total,
                used   => $used,
            };
    }
    return @res;
} # }}}

sub memory { # {{{
    my $cmd  = '/sbin/sysctl';

    my %sysctl_for = (
            'vm.stats.vm.v_page_size'      => 'pagesize',
            'vm.stats.vm.v_page_count'     => 'memsize',
            'vm.stats.vm.v_active_count'   => 'active_count',
            'vm.stats.vm.v_inactive_count' => 'inactive_count',
            'vfs.bufspace'                 => 'buffers_count',
            'vm.stats.vm.v_cache_count'    => 'cache_count',
            'vm.stats.vm.v_free_count'     => 'free_count',
        );
    my @args = (keys(%sysctl_for));
    my ($out, $err, $exit) = capture { system($cmd, @args) };
    Carp::croak "Cannot call '$cmd @args' : $err"
        if $exit;

    my @lines = split $/, $out;
    my %count_of = map {
            my ($sysctl, $val) = split(/:/, $_);
            $val =~ s/\D//g;
            ($sysctl_for{$sysctl} => $val)
        }
        @lines;

    # FIXME 20150914 geraud
    # Can't seem to find a viable way to get the swap usage via sysctl.
    my $swap_used = 0;
    foreach my $swap (_swapinfo()) {
        $swap_used += $swap->{used};
    }

    my $page_size = $count_of{pagesize};
    return (
            total    => $count_of{memsize}        * $page_size,
            active   => $count_of{active_count}   * $page_size,
            inactive => $count_of{inactive_count} * $page_size,
            cached   => $count_of{cache_count}    * $page_size,
            free     => $count_of{free_count}     * $page_size,
            swap     => $swap_used                * 1024,
            buffers  => $count_of{buffers_count},
        );
} # }}}

1;
__END__

=encoding utf-8

=head1 NAME

Sys::Stats::Freebsd - FreeBSD driver for Sys::Stats.

=head1 VERSION

This document describes Sys::Stats::Freebsd version 0.01

=head1 DESCRIPTION

You shouldn't use this module directly. Please refer to L<Sys::Stats> instead.

=head1 LICENSE

Copyright (C) Geraud CONTINSOUZAS.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Geraud CONTINSOUZAS E<lt>gcs@cpan.orgE<gt>

=cut

# vim: syn=perl nu ai cin ts=4 et sw=4 fdm=marker
